### PR TITLE
Added hook for when player takes item from container

### PR DESCRIPTION
**Pull request by cooperuser**
Added `net.minecraftforge.event.entity.player.ItemTakeEvent` that gets triggered from `net.minecraft.inventory.Slot#canTakeStack(EntityPlayer)` when a player clicks on an item.

I am currently making a mod called [InventoryFilter](https://github.com/cooper-anderson-mc/InventoryFilter) that allows the user to filter what items are allowed to go in their inventory. As of right now, I have item pickups following the filter, but I couldn't easily stop the player from grabbing from a container without checking the player's inventory every tick...

The new ItemTakeEvent event is cancelable, so it makes it easy to check what the item being clicked is, put it through the filter, and cancel if I need to.

Here is a basic example of how to use the new event:
```java
public class FilterEventHandler {
	@SubscribeEvent
	public void onItemTake(ItemTakeEvent event) {
		// Stop the player from being allowed to acquire TNT through an inventory.
		if (event.getItemStack().getItem().getRegistryName().toString().equals("minecraft:tnt")) {
			event.setCanceled(true);
		}
	}
}
```

### CODE REVIEW

File: patches/minecraft/net/minecraft/inventory/Slot.java.patch [hunk 1]
```diff
@@ -9,7 +9,18 @@
...
  
      public ItemStack func_75209_a(int p_75209_1_)
-@@ -113,4 +113,74 @@
+@@ -105,7 +105,9 @@
+ 
+     public boolean func_82869_a(EntityPlayer p_82869_1_)
+     {
+-        return true;
++        ItemStack itemStack = this.func_75211_c();
++        net.minecraft.util.EnumActionResult cancelResult = net.minecraftforge.common.ForgeHooks.onItemTake(p_82869_1_, itemStack);
```
Comment (liach) [inline patches/minecraft/net/minecraft/inventory/Slot.java.patch]: As I can see, this method is frequently overridden by mods without calling super. You might consider another place to patch so that this event remains useful.

Comment (cooperuser) [reply patches/minecraft/net/minecraft/inventory/Slot.java.patch]: Is that so... That's a shame because after looking through many classes this seemed to be the easiest implementation without modifying too many files. Based on what you said about overriding, do you think it would be acceptable if I add a new method that uses the event, and calls `canTakeStack`? I would have to go and replace all the occurrences of `canTakeStack` in Minecraft's code though, but I don't think there are too many.

Comment (cooperuser) [reply patches/minecraft/net/minecraft/inventory/Slot.java.patch]: Alternatively, I could just add the event code at each occurrence of `canTakeStack`, but I don't think that's the right way to go.

**Comment by loordgek** [comment]
you missed https://github.com/MinecraftForge/MinecraftForge/blob/1.12.x/src/main/java/net/minecraftforge/items/SlotItemHandler.java#L44

**Comment by cooperuser** [comment]
> you missed https://github.com/MinecraftForge/MinecraftForge/blob/1.12.x/src/main/java/net/minecraftforge/items/SlotItemHandler.java#L44

I see... After looking over that method it appears that I can not use it because it doesn't have access to the `entityPlayer`. I also noticed that this class has it's own `canTakeStack ` method but I don't think I can edit that one for the same reasons as the previous one.

File: patches/minecraft/net/minecraft/inventory/Slot.java.patch (hunk 1 above)
Comment (asiekierka) [reply patches/minecraft/net/minecraft/inventory/Slot.java.patch]: I don't think you should do that either.

(a) Adding a new method means you're incompatible with all inventories which call canTakeStack on their own.
(b) Replacing each occurence won't cover mods.

I think that you should do a clean implementation and simply wait for 1.13 with it. There's no way to do this cleanly in 1.12 this late into the game.

**Comment by loordgek** [comment]
srry wrong one https://github.com/MinecraftForge/MinecraftForge/blob/1.12.x/src/main/java/net/minecraftforge/items/SlotItemHandler.java#L127

**Comment by cooperuser** [comment]
I've been looking over that and reading some documentation and I am not quite sure how I actually use that function. Is it something to do with Capabilities? As far as I can see `SlotItemHandler` is not used anywhere and I'm not sure how to get it to override the base minecraft `Slot` class. Am I missing something obvious?

File: patches/minecraft/net/minecraft/inventory/Slot.java.patch (hunk 1 above)
Comment (cooperuser) [reply patches/minecraft/net/minecraft/inventory/Slot.java.patch]: Hmm... That's a shame. For now do you think I should temporarily do something in my own mod as some sort of workaround for this?

**Comment by raoulvdberge** [comment]
SlotItemHandler is used by mods who use IItemHandler inventories, which are the vast majority of mods in 1.12.2.

**Comment by cooperuser** [comment]
I see. So should the event just be triggered from both versions of `canTakeItem`? That way it will work with vanilla and modded inventories.

File: src/main/java/net/minecraftforge/items/SlotItemHandler.java [hunk 2]
```diff
@@ -126,7 +126,8 @@ public int getItemStackLimit(@Nonnull ItemStack stack)
     @Override
     public boolean canTakeStack(EntityPlayer playerIn)
     {
-        return !this.getItemHandler().extractItem(index, 1, true).isEmpty();
+        net.minecraft.util.EnumActionResult cancelResult = net.minecraftforge.common.ForgeHooks.onItemTake(playerIn, this.getStack());
```
Comment (cooperuser) [inline src/main/java/net/minecraftforge/items/SlotItemHandler.java]: I added the event triggering code to `SlotItemHandler#canTakeStack()`. If I understood @liach and @raoulvdberge correctly, this should do the trick.

**Comment by cooperuser** [comment]
Are there any changes that need to be made to this?

**Review by liach — APPROVED** [review]
This should be fine enough.

**Comment by BlayTheNinth** [comment]
The name is misleading, it implies an item is actually being taken when `canTakeStack` just determines if it's possible at all. Vanilla anvils call it from their GUI rendering for example. Maybe should keep the "Can" in the event name to prevent that confusion.

Also, this should probably be using `@HasResult` on the event instead of a seperate `EnumActionResult` field. EnumActionResult is generally used for item / block interactions, not for event results.

**Comment by cooperuser** [comment]
Ah, that's a good point about the name being misleading. I guess `onCanMoveStack` would work better in that case? I just looked into how it it is used in anvils, and it appears to be used for the text color? I'm going to do a test to see if causes some unintended things to happen.

**Comment by cooperuser** [comment]
I am also currently looking into switching to `@HasResult`

**Comment by cooperuser** [comment]
It appears the events using `@HasResult` are in `ForgeEventFactory` instead of `ForgeHooks`. Also it appears the naming convention there would make it `onCanItemMove`? maybe `onCanStackMove`?

**Comment by cooperuser** [comment]
Okay I implemented the changes stated.
Here is a new version of the example mod because the name of the event changed:

```java
public class FilterEventHandler {
	@SubscribeEvent
	public void onItemTake(CanTakeItemEvent event) {
		// Stop the player from being allowed to acquire TNT through an inventory.
		if (event.getItemStack().getItem().getRegistryName().toString().equals("minecraft:tnt")) {
			event.setCanceled(true);
		}
	}
}
```

**Comment by kierajreed** [comment]
You need to add a license header to `CanTakeItemEvent.java`... otherwise the Travis build will not pass because the license isn't there.

**Comment by cooperuser** [comment]
Adding that now.  I can just copy it from a different file and paste it there right?

**Comment by cooperuser** [comment]
Also I'm pretty sure I did not have a license header in the file before I renamed it... which is strange because it did not error then

**Comment by liach** [comment]
Do `gradle licenseFormat` to fix the license.

**Comment by cooperuser** [comment]
`gradle licenseFormat` didn't appear to do anything, or at least no changes that I could commit to fix the problem, so I added the license manually.

**Comment by cooperuser** [comment]
I'm just wondering why I couldn't get the license error to show up on my machine

**Comment by liach** [comment]
@cooper-anderson You need to do `gradle checkLicenses` to check licenses.

**Comment by cooperuser** [comment]
I just tried running that command both with and without the license header and it passed both times

**Comment by cooperuser** [comment]
image

**Comment by mezz** [comment]
The license check is passing as of your last commit, you don't need to worry about it any more.

**Comment by cooperuser** [comment]
Sounds good. I have been reading through [Minecrell/licenser](https://github.com/Minecrell/licenser) to see if I can find anything that might be causing this to happen.

**Comment by kierajreed** [comment]
There was a bad license in an earlier PR that was merged and later caused several other PRs' builds to fail.

**Comment by cooperuser** [comment]
Ah, thanks for clearing that up!

**Comment by cooperuser** [comment]
Now that all that has been sorted out, is there anything else that needs be done to this PR?

File: src/main/java/net/minecraftforge/event/ForgeEventFactory.java [hunk 3]
```diff
@@ -489,6 +489,13 @@ public static int onItemPickup(EntityItem entityItem, EntityPlayer player)
         return event.getResult() == Result.ALLOW ? 1 : 0;
     }
 
+    public static boolean onCanTakeItem(EntityPlayer player, ItemStack itemStack)
```
Comment (mezz) [inline src/main/java/net/minecraftforge/event/ForgeEventFactory.java]: Why is this here?

File: src/main/java/net/minecraftforge/items/SlotItemHandler.java [hunk 4]
```diff
@@ -126,7 +126,8 @@ public int getItemStackLimit(@Nonnull ItemStack stack)
     @Override
     public boolean canTakeStack(EntityPlayer playerIn)
     {
-        return !this.getItemHandler().extractItem(index, 1, true).isEmpty();
+        boolean result = net.minecraftforge.event.ForgeEventFactory.onCanTakeItem(playerIn, this.getStack());
+        return result && !this.getItemHandler().extractItem(index, 1, true).isEmpty();
```
Comment (mezz) [inline src/main/java/net/minecraftforge/items/SlotItemHandler.java]: Would it be useful for the event to have a "vanilla result" it can reference?

**Review by mezz — CHANGES_REQUESTED** [review]

File: src/main/java/net/minecraftforge/items/SlotItemHandler.java (hunk 4 above)
Comment (cooperuser) [reply src/main/java/net/minecraftforge/items/SlotItemHandler.java]: What do you mean exactly? Would this be something that can be accessed by an instance of the event?

Comment (Alpvax) [reply src/main/java/net/minecraftforge/items/SlotItemHandler.java]: Passing the result of `!this.getItemHandler().extractItem(index, 1, true).isEmpty()` to the event, so that modders can see whether players could take the item in vanilla.

Comment (cooperuser) [reply src/main/java/net/minecraftforge/items/SlotItemHandler.java]: I see. That could be useful. Is there already a convention used in other events for what this method should be called?

Comment (cooperuser) [reply src/main/java/net/minecraftforge/items/SlotItemHandler.java]: Also, the `canTakeStack()` method in the vanilla `Slot` class always returns true. Should we just pass true into the event? or should we add the check manually.

Comment (Alpvax) [reply src/main/java/net/minecraftforge/items/SlotItemHandler.java]: Add the check. That way if vanilla changes something down the line, we don't have to.

Comment (cooperuser) [reply src/main/java/net/minecraftforge/items/SlotItemHandler.java]: Good point. Now for the name, I was thinking `getAllowed()` or something along those lines. `getVanillaResult()` would also work

Comment (cooperuser) [reply src/main/java/net/minecraftforge/items/SlotItemHandler.java]: I implemented the check in `SlotItemHandler`, but I was unable to in `Slot` because `canTakeStack()` doesn't have the `ItemStack` as a parameter, and adding that would require editing lots of minecraft files and I thought it would be best to avoid that.

File: src/main/java/net/minecraftforge/items/SlotItemHandler.java [hunk 5]
```diff
@@ -126,7 +126,9 @@ public int getItemStackLimit(@Nonnull ItemStack stack)
     @Override
     public boolean canTakeStack(EntityPlayer playerIn)
     {
-        return !this.getItemHandler().extractItem(index, 1, true).isEmpty();
+        boolean vanillaResult = !this.getItemHandler().extractItem(index, 1, true).isEmpty();
+        boolean result = net.minecraftforge.event.ForgeEventFactory.onCanTakeItem(playerIn, this.getStack(), vanillaResult);
+        return result && vanillaResult;
```
Comment (mezz) [inline src/main/java/net/minecraftforge/items/SlotItemHandler.java]: If `vanillaResult` here is `false`, is there a point in firing the event?
I think we should carefully consider if this is the right logic, and if it is we should make it something that avoids firing when it's not required, like
`return vanillaResult && net.minecraftforge.event.ForgeEventFactory.onCanTakeItem(playerIn, this.getStack(), vanillaResult);`

**Review by mezz — CHANGES_REQUESTED** [review]

File: src/main/java/net/minecraftforge/items/SlotItemHandler.java (hunk 5 above)
Comment (cooperuser) [reply src/main/java/net/minecraftforge/items/SlotItemHandler.java]: I thought about that too. But then it occurred to me that a mod maker might want to override the `false` vanilla result and force the item to be picked up. In that case should we just remove `vanillaResult` from the return statement? If not then I think making the event only trigger when it is true also works.

Comment (cooperuser) [reply src/main/java/net/minecraftforge/items/SlotItemHandler.java]: Any thoughts?

Comment (mezz) [reply src/main/java/net/minecraftforge/items/SlotItemHandler.java]: Let's make it so that the event result wins on this implementation. It makes it more flexible so that events can control it, but still allows mods to override it. So yeah, just remove `vanillaResult` from the return statement

Comment (cooperuser) [reply src/main/java/net/minecraftforge/items/SlotItemHandler.java]: Sounds good. Besides removing `vanillaResult` from that return statement, is there anything else that we need to do?

File: src/main/java/net/minecraftforge/event/ForgeEventFactory.java [hunk 6]
```diff
@@ -489,6 +489,13 @@ public static int onItemPickup(EntityItem entityItem, EntityPlayer player)
         return event.getResult() == Result.ALLOW ? 1 : 0;
     }
 
+    public static boolean onCanTakeItem(EntityPlayer player, ItemStack itemStack, boolean vanillaResult)
+    {
+        Event event = new net.minecraftforge.event.entity.player.CanTakeItemEvent(itemStack, player, vanillaResult);
+        MinecraftForge.EVENT_BUS.post(event);
+        return !event.isCanceled() && event.getResult() != Result.DENY;
```
Comment (mezz) [inline src/main/java/net/minecraftforge/event/ForgeEventFactory.java]: I think this event should only use `Result` and not canceled, so that `Result.DEFAULT` means use `vanillaResult`, that way implementations like in `SlotItemHandler` are still correct.
With the way it's implemented now, `SlotItemHandler` will allow taking items by default, which changes how it works.

**Review by mezz — CHANGES_REQUESTED** [review]

File: src/main/java/net/minecraftforge/event/ForgeEventFactory.java (hunk 6 above)
Comment (cooperuser) [reply src/main/java/net/minecraftforge/event/ForgeEventFactory.java]: So just remove the `!event.isCanceled() && ` part? or did I understand that wrong

Comment (mezz) [reply src/main/java/net/minecraftforge/event/ForgeEventFactory.java]: we need 3 cases:
* `Result.DENY`: return `false`
* `Result.DEFAULT`: return `vanillaResult`
* `Result.ALLOW`: return `true`

If you look at `SlotItemHandler` you can see why. Imagine nobody is listening to the event, this code is changing the method's behavior unless the `DEFAULT` action returns `vanillaResult`

Canceled is not really needed here as a way of changing the result, it's just to keep other mods from altering the result later.

Comment (cooperuser) [reply src/main/java/net/minecraftforge/event/ForgeEventFactory.java]: I see... So should I implement these all on one line? or have an if statement with 3 separate returns. So far I have been using one line returns, but this seems a bit complicated to read for anyone looking over the code

Comment (mezz) [reply src/main/java/net/minecraftforge/event/ForgeEventFactory.java]: Forge files like `ForgeEventFactory` aren't patches so they do not have line restrictions, so you can write it in a way that is more readable.

Comment (cooperuser) [reply src/main/java/net/minecraftforge/event/ForgeEventFactory.java]: Okay I did it using if statements for readability. Also, I put `return vanillaResult` as the default after the other cases.

File: src/main/java/net/minecraftforge/event/ForgeEventFactory.java [hunk 7]
```diff
@@ -489,6 +489,18 @@ public static int onItemPickup(EntityItem entityItem, EntityPlayer player)
         return event.getResult() == Result.ALLOW ? 1 : 0;
     }
 
+    public static boolean onCanTakeItem(EntityPlayer player, ItemStack itemStack, boolean vanillaResult)
+    {
+        Event event = new net.minecraftforge.event.entity.player.CanTakeItemEvent(itemStack, player, vanillaResult);
+        MinecraftForge.EVENT_BUS.post(event);
+        if (event.getResult() == Result.DENY) {
+            return false;
+        } else if (event.getResult() == Result.ALLOW) {
```
Comment (mezz) [inline src/main/java/net/minecraftforge/event/ForgeEventFactory.java]: Please follow Forge's brace style. You can use no braces for short statements like this, or put braces on a new line.

**Review by mezz — CHANGES_REQUESTED** [review]

**Review by mezz — APPROVED** [review]

File: patches/minecraft/net/minecraft/inventory/Slot.java.patch [hunk 8]
```diff
@@ -9,7 +9,17 @@
...
  
      public ItemStack func_75209_a(int p_75209_1_)
-@@ -113,4 +113,74 @@
+@@ -105,7 +105,8 @@
+ 
+     public boolean func_82869_a(EntityPlayer p_82869_1_)
+     {
+-        return true;
++        ItemStack itemStack = this.func_75211_c();
++        return net.minecraftforge.event.ForgeEventFactory.onCanTakeItem(p_82869_1_, itemStack, true);
```
Comment (LexManos) [inline patches/minecraft/net/minecraft/inventory/Slot.java.patch]: Why is this two lines?

File: src/main/java/net/minecraftforge/event/entity/player/CanTakeItemEvent.java [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+ * The event can be canceled, and no further processing will be done.
+ *
+ * This event is fired in {@link net.minecraft.inventory.Slot#canTakeStack(EntityPlayer)} and {@link net.minecraftforge.items.SlotItemHandler#canTakeStack(EntityPlayer)} when a player clicks on an item.
+ */
+@Cancelable
+@Event.HasResult
+public class CanTakeItemEvent extends PlayerEvent
+{
+    private final ItemStack itemStack;
+    private final boolean vanillaResult;
```
Comment (LexManos) [inline src/main/java/net/minecraftforge/event/entity/player/CanTakeItemEvent.java]: So absolutly nothing else is useful in this event? Not the inventory, or the slot, or anything like that?

**Review by LexManos — DISMISSED** [review]

**Comment by cooper-anderson** [comment]
Is there anything else that needs to be done here?

**Comment by cooperuser** [comment]
Is there anything else that needs to be done here?

File: src/main/java/net/minecraftforge/items/SlotItemHandler.java [hunk 10]
```diff
@@ -126,7 +126,8 @@ public int getItemStackLimit(@Nonnull ItemStack stack)
     @Override
     public boolean canTakeStack(EntityPlayer playerIn)
     {
-        return !this.getItemHandler().extractItem(index, 1, true).isEmpty();
+        boolean vanillaResult = !this.getItemHandler().extractItem(index, 1, true).isEmpty();
+        return net.minecraftforge.event.ForgeEventFactory.onCanTakeItem(this.getStack(), this, this.inventory, playerIn, vanillaResult);
```
Comment (tterrag1098) [inline src/main/java/net/minecraftforge/items/SlotItemHandler.java:130]: The `inventory` field is meaningless in this class. You need to provide the `IItemHandler` to the event somehow. In fact perhaps the event should *only* accept `IItemHandler` and be wrapped for vanilla cases.

File: src/main/java/net/minecraftforge/event/entity/player/CanTakeItemEvent.java [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+import net.minecraftforge.fml.common.eventhandler.Event;
+import javax.annotation.Nonnull;
+
+/**
+ * This event is called when a player clicks on an item in a container.
+ * The event can be canceled, and no further processing will be done.
+ *
+ * This event is fired in {@link net.minecraft.inventory.Slot#canTakeStack(EntityPlayer)} and {@link net.minecraftforge.items.SlotItemHandler#canTakeStack(EntityPlayer)} when a player clicks on an item.
+ */
+@Cancelable
```
Comment (tterrag1098) [inline src/main/java/net/minecraftforge/event/entity/player/CanTakeItemEvent.java:36]: No point having this if there is a result. You don't even use the canceled state in the event post code.

**Review by tterrag1098 — CHANGES_REQUESTED** [review]
